src/OVAL/oval_session.c

```diff
@@ -0,0 +1,463 @@
+		return ret;
+	}
+	if ((ret = oval_session_load_variables(session)) != 0) {
+		return ret;
+	}
+
+	return ret;
+}
+
+static int oval_session_setup_agent(struct oval_session *session)
+{
+	__attribute__nonnull__(session);
+
+	char *path_clone;
+
+	path_clone = oscap_strdup(oscap_source_readable_origin(session->oval.definitions));
+	if (path_clone == NULL) {
+		return 1;
+	}
+
+	/* free the previous Agent session if this function was already called */
+	if (session->sess)
+		oval_agent_destroy_session(session->sess);
+
+	session->sess = oval_agent_new_session(session->def_model, basename(path_clone));
+	if (session->sess == NULL) {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "Failed to create a new agent session.");
+		oscap_free(path_clone);
+		return 1;
+	}
+	oscap_free(path_clone);
+
+	oval_agent_set_product_name(session->sess, (char *)oscap_productname);
+	return 0;
+}
+
+int oval_session_evaluate_id(struct oval_session *session, char *probe_root, const char *id, oval_result_t *result)
+{
+	__attribute__nonnull__(session);
+
+	if (probe_root) {
+		if (setenv("OSCAP_PROBE_ROOT", probe_root, 1) != 0) {
+			oscap_seterr(OSCAP_EFAMILY_OSCAP, "Failed to set the OSCAP_PROBE_ROOT environment variable.");
+			return 1;
+		}
+	}
+
+	if (id == NULL) {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "No OVAL Definion id set.");
+		return 1;
+	}
+
+	if (oval_session_setup_agent(session) != 0) {
+		return 1;
+	}
+
+	oval_agent_eval_definition(session->sess, id);
+	*result = OVAL_RESULT_NOT_EVALUATED;
+	oval_agent_get_definition_result(session->sess, id, result);
+	if (oscap_err()) {
+		return 1;
+	}
+
+	session->res_model = oval_agent_get_results_model(session->sess);
+
+	return 0;
+}
+
+int oval_session_evaluate(struct oval_session *session, char *probe_root, agent_reporter fn, void *arg)
+{
+	__attribute__nonnull__(session);
+
+	if (probe_root) {
+		if (setenv("OSCAP_PROBE_ROOT", probe_root, 1) != 0) {
+			oscap_seterr(OSCAP_EFAMILY_OSCAP, "Failed to set the OSCAP_PROBE_ROOT environment variable.");
+			return 1;
+		}
+	}
+
+	if (oval_session_setup_agent(session) != 0) {
+		return 1;
+	}
+
+	oval_agent_eval_system(session->sess, fn, arg);
+	if (oscap_err()) {
+		return 1;
+	}
+
+	session->res_model = oval_agent_get_results_model(session->sess);
+
+	return 0;
+}
+
+int oval_session_export(struct oval_session *session)
+{
+	__attribute__nonnull__(session);
+
+	struct oval_directives_model *dir_model = NULL;
+	struct oscap_source *result = NULL;		/* OVAL Results */
+	const char *filename = NULL;
+	int ret = 0;
+
+	/* Import OVAL Directives if any */
+	if (session->oval.directives && session->res_model) {
+		/* OVAL Directives can be used only if there are OVAL Resutls and these are
+		 * only available if there is a Results model which mean that either
+		 * evaluation or analyse was performed. */
+		if (dir_model)
+			oval_directives_model_free(dir_model);
+		oval_directives_model_import_source(dir_model, session->oval.directives);
```

Comment:
What about checking return value -- directives model may be malformed.

(perhaps this has been always broken, not sure)